### PR TITLE
Add `FileMap` options to `put_filemap()` Fix function.

### LINKED ISSUE #265: Add options for multi column lookup to `put_filemap()`
@blackwinter Introduced multi column lookup to metafacture morph. to use this in fix we need some additional options with `put_filemap()`

https://github.com/metafacture/metafacture-core/pull/471

### PR DESCRIPTION
**Pull request by blackwinter**
WIP until a stable metafacture-core release artifact becomes available.

Resolves #265.

### CODE REVIEW

**Comment by blackwinter** [comment]
There seems to be an issue with the metafactore-core artifacts. All integration tests are failing:

```
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/home/wille/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-log4j12/1.7.21/7238b064d1aba20da2ac03217d700d91e02460fa/slf4j-log4j12-1.7.21.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/home/wille/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-simple/1.7.21/be4b3c560a37e69b6c58278116740db28832232c/slf4j-simple-1.7.21.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
Exception in thread "main" org.metafacture.commons.reflection.ReflectionException: class could not be instantiated: class org.metafacture.metafix.Metafix
        at org.metafacture.commons.reflection.ConfigurableClass.newInstance(ConfigurableClass.java:149)
        at org.metafacture.commons.reflection.ObjectFactory.newInstance(ObjectFactory.java:105)
        at org.metafacture.flux.parser.FluxProgramm.createElement(FluxProgramm.java:78)
        at org.metafacture.flux.parser.FluxProgramm.addElement(FluxProgramm.java:90)
        at org.metafacture.flux.parser.FlowBuilder.pipe(FlowBuilder.java:736)
        at org.metafacture.flux.parser.FlowBuilder.flowtail(FlowBuilder.java:514)
        at org.metafacture.flux.parser.FlowBuilder.flow(FlowBuilder.java:226)
        at org.metafacture.flux.parser.FlowBuilder.flux(FlowBuilder.java:122)
        at org.metafacture.flux.FluxCompiler.compileFlow(FluxCompiler.java:66)
        at org.metafacture.flux.FluxCompiler.compile(FluxCompiler.java:54)
        at org.metafacture.runner.Flux.main(Flux.java:85)
Caused by: java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at org.metafacture.commons.reflection.ConfigurableClass.newInstance(ConfigurableClass.java:144)
        ... 10 more
Caused by: java.lang.NoSuchFieldError: EOF_TOKEN
        at org.eclipse.xtext.parser.antlr.Lexer.nextToken(Lexer.java:60)
        at org.antlr.runtime.BufferedTokenStream.fetch(BufferedTokenStream.java:143)
        at org.antlr.runtime.BufferedTokenStream.sync(BufferedTokenStream.java:137)
        at org.antlr.runtime.BufferedTokenStream.fill(BufferedTokenStream.java:286)
        at org.antlr.runtime.BufferedTokenStream.toString(BufferedTokenStream.java:251)
        at org.eclipse.xtext.parser.antlr.AbstractInternalAntlrParser.parse(AbstractInternalAntlrParser.java:586)
        at org.eclipse.xtext.parser.antlr.AbstractAntlrParser.doParse(AbstractAntlrParser.java:103)
        at org.eclipse.xtext.parser.antlr.AbstractAntlrParser.parse(AbstractAntlrParser.java:85)
        at org.eclipse.xtext.parser.antlr.AbstractAntlrParser.doParse(AbstractAntlrParser.java:63)
        at org.eclipse.xtext.parser.AbstractParser.parse(AbstractParser.java:34)
        at org.eclipse.xtext.resource.XtextResource.doLoad(XtextResource.java:178)
        at org.eclipse.xtext.linking.lazy.LazyLinkingResource.doLoad(LazyLinkingResource.java:115)
        at org.eclipse.emf.ecore.resource.impl.ResourceImpl.load(ResourceImpl.java:1563)
        at org.eclipse.emf.ecore.resource.impl.ResourceImpl.load(ResourceImpl.java:1342)
        at org.eclipse.xtext.resource.persistence.StorageAwareResource.load(StorageAwareResource.java:64)
        at org.eclipse.emf.ecore.resource.impl.ResourceSetImpl.demandLoad(ResourceSetImpl.java:259)
        at org.eclipse.emf.ecore.resource.impl.ResourceSetImpl.demandLoadHelper(ResourceSetImpl.java:274)
        at org.eclipse.xtext.resource.XtextResourceSet.getResource(XtextResourceSet.java:266)
        at org.eclipse.xtext.resource.SynchronizedXtextResourceSet.getResource(SynchronizedXtextResourceSet.java:33)
        at org.metafacture.metafix.validation.XtextValidator.getResource(XtextValidator.java:63)
        at org.metafacture.metafix.validation.XtextValidator.getValidatedResource(XtextValidator.java:67)
        at org.metafacture.metafix.FixStandaloneSetup.parseFix(FixStandaloneSetup.java:31)
        at org.metafacture.metafix.Metafix.lambda$getRecordTransformer$0(Metafix.java:162)
        at java.util.HashMap.computeIfAbsent(HashMap.java:1128)
        at org.metafacture.metafix.Metafix.getRecordTransformer(Metafix.java:162)
        at org.metafacture.metafix.Metafix.<init>(Metafix.java:111)
        ... 15 more

FAILURE: Build failed with an exception.
```

**Comment by blackwinter** [comment]
Note that we've seen this error before in #90.

**Comment by blackwinter** [comment]
> Seems similar to [hbz/lobid-resources#1462](https://github.com/hbz/lobid-resources/issues/1462) - a conflict between different `xtext` libraries.

How do you figure?

**Comment by dr0i** [comment]
Because of your error message (how did you get this , btw - I only see that it fails). It's said there "NoSuchFieldError" and that reminded me. Also, locally build and consumed mf-master works, and `metafacture-flux` (which uses ANTLR) has the same bytes on github-packages as the locally one. But this is only a "smell", no really proof, so this may be misleading.

**Comment by blackwinter** [comment]
> It's said there "NoSuchFieldError" and that reminded me.

True, but it's a different field. `EOF_TOKEN` comes from ANTLR, not Xtext. As I said, it's the same error as in #90.

**Comment by blackwinter** [comment]
> how did you get this , btw - I only see that it fails

Ran it locally. Maybe we should change the integration tests to print errors to the console in CI.

**Comment by dr0i** [comment]
Yeah but xtext uses ANTLR. I am unsure - is there no connection?

**Comment by blackwinter** [comment]
> Yeah but xtext uses ANTLR. I am unsure - is there no connection?

I don't know. Since we weren't able to identify the actual root cause in hbz/lobid-resources#1462 or any cause in #90, I really can't say. It's clearly reminiscent of the former (works in one environment, but not the other), so it might make sense to investigate in that direction.

**Comment by blackwinter** [comment]
> Maybe we should change the integration tests to print errors to the console in CI.

Done in 9ee2194.

**Comment by dr0i** [comment]
@blackwinter : updated to a "stable" `5.4.1-rc1` and force psuhed. Rebased. Fixed the integration test error. You may want take over?

**Comment by blackwinter** [comment]
Thanks @dr0i. I can't say I particularly like the version pinning, but let's just move on I guess...

Ready for review.

**Review by dr0i — APPROVED** [review]

**Comment by dr0i** [comment]
Thx for correcting the commit.

>  I can't say I particularly like the version pinning

Why?The underlying problem seems to be a conflicting dependency , i.e. the erratic choosing of an one of the two `antrl` libraries in the classpath. Pinning the correct version in `metafix-runner` makes this a stable solution, no?  Resp. is there a better solution?

**Comment by blackwinter** [comment]
Because it's non-obvious (to me, at least). No dependencies have changed between 5.4.0 and 5.4.1-rc1, only the publishing process. I don't see anything erratic about it either, it's consistently selecting the wrong version - but only in the runner (integration tests), not in the library (unit tests). I don't understand it, pure and simple.

**Comment by dr0i** [comment]
From https://stackoverflow.com/questions/24962607/multiple-versions-of-the-same-dependency-in-maven :
>  Note that if two dependency versions are at the same depth in the dependency tree, the first declaration wins.

My understanding (may be wrong - but "explains" the phenomenon):
What _has_ changed is the packaging (done as GitHub packages). Assumption: The build of these packages changes 
the order of the dependency tree (_why_ and _how_  I also don't understand, admittedly). Thus, consuming the GitHub packages results in picking versionA while in local builds versionB is choosen.

**Comment by blackwinter** [comment]
> Thus, consuming the GitHub packages results in picking versionA while in local builds versionB is choosen.

Both `metafix` (unit tests) and `metafix-runner` (integration tests) consume the same package do they not? And integration tests fail(ed) both locally and in CI.

**Comment by dr0i** [comment]
Oh, yes, you are correct.
So, let's compare the dependency-tree:
```
git reset --hard 4976755d3a2182d71059c38ea34fdab5347bc4ec # where everything was fine
./gradlew metafix-runner:dependencies > before266_allfine.txt
git reset --hard 6d82e08df665ffd5b90a8aba86dc5aa446267c63 # where things broke
./gradlew metafix-runner:dependencies > 255_broken.txt
diff 255_broken.txt before266_allfine.txt
```
in 255_broken.txt there is:
> [...]
runtimeClasspath - Runtime classpath of source set 'main'.
+--- org.eclipse.xtext:xtext-dev-bom:2.26.0
[...]
|    +--- org.antlr:antlr-runtime:3.2 -> 3.5.2 (c)
[...]

while in before266_allfine.txt there is:
> |    +--- org.antlr:antlr-runtime:3.2 (c)

So: why is there "3.5.2 (c)" using GithubPackages?

**Comment by dr0i** [comment]
@blackwinter can we merge this and resolve https://github.com/metafacture/metafacture-fix/issues/265 ?
We could further investigate the version problem here or open another issue, if you want.

**Comment by blackwinter** [comment]
> can we merge this and resolve #265 ?

@TobiasNx hasn't finished his functional review.

**Comment by blackwinter** [comment]
> We could further investigate the version problem here or open another issue, if you want.

Yes. We should probably do it in a dedicated issue, but I'm not sure what the right angle would be (I so desperately want Metafacture's Gradle build cleaned up, and also updated to a current Gradle version). And seeing as we already started here, let me just post another observation: The new publishing process (via `maven-publish` plugin, e.g. `publishToMavenLocal`) includes metafacture-flux's antlr dependency (via `antlr` plugin), while the old process (via `maven` plugin, e.g. `uploadArchives`) did not.

```diff
--- metafacture-flux-uploadArchives-pom.xml 2022-11-15 14:39:31.375148909 +0100
+++ metafacture-flux-publishToMavenLocal-pom.xml  2022-11-15 14:39:15.143648173 +0100
@@ -63,6 +68,12 @@
     <url>https://github.com/metafacture/metafacture-core/actions</url>
   </ciManagement>
   <dependencies>
+    <dependency>
+      <groupId>org.antlr</groupId>
+      <artifactId>antlr</artifactId>
+      <version>3.5.2</version>
+      <scope>compile</scope>
+    </dependency>
     <dependency>
       <groupId>org.metafacture</groupId>
       <artifactId>metafacture-framework</artifactId>
@@ -81,17 +92,5 @@
       <version>master-SNAPSHOT</version>
       <scope>runtime</scope>
     </dependency>
-    <dependency>
-      <groupId>junit</groupId>
-      <artifactId>junit</artifactId>
-      <version>4.12</version>
-      <scope>test</scope>
-    </dependency>
-    <dependency>
-      <groupId>org.metafacture</groupId>
-      <artifactId>metafacture-plumbing</artifactId>
-      <version>master-SNAPSHOT</version>
-      <scope>test</scope>
-    </dependency>
   </dependencies>
 </project>
```

**Comment by dr0i** [comment]
>>>    Thus, consuming the GitHub packages results in picking versionA while in local builds versionB is choosen.

>>Both metafix (unit tests) and metafix-runner (integration tests) consume the same package do they not? And integration tests fail(ed) both locally and in CI.

>Oh, yes, you are correct.

Wait ! What I meant is:
Building `metafacture-core` locally and consuming this build being on  6d82e08df665ffd5b90a8aba86dc5aa446267c63 (i.e. without pinning the antlr version) _does_ make `metafix` build sucessfully (and integration tests pass). It fails _only_ when using the GitHub packages.

Locally building `metafacture-core:metafacture-runner` shows that the dependency is `org.antlr:antlr:3.5.2` (coming from `metafacture-flux`) _but_ when consuming `metafacture-runner` the `antlr` version is `3.2` (somehow ignoring the `antlr` version from `metafacture-flux`).

Looking into the different dependencies:
The failing (aka GithubPackages) one has:
>
|    +--- org.metafacture:metafacture-flux:metafacture-core-5.4.1-rc1 
|    |    +--- org.metafacture:metafacture-commons:metafacture-core-5.4.1-rc1
|    |    +--- org.metafacture:metafacture-io:metafacture-core-5.4.1-rc1 (*)
|    |    +--- org.antlr:antlr:3.5.2

(i.e. `metafacture-flux` brings `org.antlr:antlr:3.5.2`)

while the locally consumed one lacks that :
>
+--- org.metafacture:metafacture-runner:master-SNAPSHOT
|    +--- org.metafacture:metafacture-framework:master-SNAPSHOT
|    +--- org.metafacture:metafacture-commons:master-SNAPSHOT
|    \--- org.metafacture:metafacture-flux:master-SNAPSHOT
|         +--- org.metafacture:metafacture-framework:master-SNAPSHOT
|         +--- org.metafacture:metafacture-commons:master-SNAPSHOT
|         \--- org.metafacture:metafacture-io:master-SNAPSHOT (*)

Idk why. :confused: .

**Comment by blackwinter** [comment]
> Ah, that explains it!

Well, not exactly. We still don't know _why_... (need to look into the `antlr` plugin next; maybe some dependency mapping issue: [maven](https://docs.gradle.org/6.0/userguide/maven_plugin.html#sub:dependency_mapping) vs. [maven-publish](https://docs.gradle.org/current/userguide/publishing_maven.html#publishing_maven:resolved_dependencies)).

File: metafix/src/test/resources/org/metafacture/metafix/integration/lookup/fromJson/toJson/lookupSimpleWithMultiColumnFile/test.fix [hunk 1]
```diff
@@ -0,0 +1,6 @@
+put_filemap("idLookup", "./mapfile.tsv", sep_char:"\t",key_column:"1",value_column:"2",expected_columns:"3")
+put_filemap("nameLookup", "./mapfile.tsv", sep_char:"\t",expected_columns:"3")
```
Comment (blackwinter) [inline metafix/src/test/resources/org/metafacture/metafix/integration/lookup/fromJson/toJson/lookupSimpleWithMultiColumnFile/test.fix:2]: You've got the arguments reversed: `put_filemap("<sourceFile>", "<mapName>", ...)`

Comment (TobiasNx) [reply metafix/src/test/resources/org/metafacture/metafix/integration/lookup/fromJson/toJson/lookupSimpleWithMultiColumnFile/test.fix]: Thanks.

**Comment by TobiasNx** [comment]
There are options that are unknown to me. I could add a documentation, if somebody explains their function to me.

`compression`
`decompress_concatenated`
`encoding`
`ignore_pattern`

**Comment by blackwinter** [comment]
All options are documented in the [source](https://github.com/metafacture/metafacture-core/blob/6e955f75bde208ae7aad0613244e52745868232f/metamorph/src/main/java/org/metafacture/metamorph/maps/FileMap.java) ([Javadoc](https://javadoc.io/static/org.metafacture/metamorph/5.4.0/org/metafacture/metamorph/maps/FileMap.html); `ignore_pattern` hasn't been released yet, though).

**Comment by blackwinter** [comment]
Please note that I already added some `lookup()` documentation in 7701a9e.

**Comment by TobiasNx** [comment]
Otherwise +1 from me concerning functional review.

**Comment by blackwinter** [comment]
> I added some documentation. [...] Otherwise +1 from me concerning functional review.

Thanks! Can you approve then?

File: README.md [hunk 2]
```diff
@@ -190,6 +190,16 @@ The separator (`sep_char`) will vary depending on the source file, e.g.:
...
 
+Additional options:
+
+- `key_column` defines the column to be used for keys. Uses zero index. Default value: `0`.
+- `value_column` defines the column to be used for values. Uses zero index. Default value: `1`.
+- `expected_columns` sets number of expected columns; lines with different number of columns are ignored. Set to `-1` to disable the check and allow arbitrary number of columns. Default value: `2`.
+- `allow_empty_values` sets whether to allow empty values in the filemap or to ignore these entries. Default value: `false`.
+- `compression` sets the compression of the file.
+- `decompress_concatenated` flags whether to use decompress concatenated file compression.
+- `encoding` sets the encoding used to open the resource.
```
[the remark this answers is not in the document]
Comment (blackwinter) [reply README.md:201]: Oh, and also alphabetical order.

**Pull request merged**